src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/IRequestExecutionQueue.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+#nullable enable
+
+namespace CommonLanguageServerProtocol.Framework;
+
+public interface IRequestExecutionQueue<RequestContextType>
+{
+    Task<TResponseType> ExecuteAsync<TRequestType, TResponseType>(bool mutatesSolutionState, bool requiresLSPSolution, IRequestHandler<TRequestType, TResponseType, RequestContextType> handler, TRequestType? request, string methodName, ILspServices lspServices, CancellationToken cancellationToken);
+
+    Task ExecuteAsync<TRequestType>(bool mutatesSolutionState, bool requiresLSPSolution, INotificationHandler<TRequestType, RequestContextType> handler, TRequestType? request, string methodName, ILspServices lspServices, CancellationToken cancellationToken);
+
+    Task ExecuteAsync(bool mutatesSolutionState, bool requiresLSPSolution, INotificationHandler<RequestContextType> handler, string methodName, ILspServices lspServices, CancellationToken cancellationToken);
+
+    void Start(ILspServices lspServices);
+
+    void Shutdown();
+
+    event EventHandler<RequestShutdownEventArgs>? RequestServerShutdown;
```

Comment:
I'd be curious on other folks view of `event EventHandler<T>` vs. `event Action<T>` I feel like the EventHandler bits are usually heavy handed is all

Reply:
what is this event used/needed for?

Reply:
>  I feel like the EventHandler bits are usually heavy handed is all

I don't personally have strong feelings either way, just used what was already in Roslyn.

> what is this event used/needed for?

It allows the LanguageServer to [know when it ought to shut down](https://github.com/dotnet/roslyn/pull/62924/files#diff-68827171a9800c8a40f4ccd71a96a93fb5206420213e3532846c7ab0fa0f453fR151) because the queue is broken.

Reply:
I personally don't like the events - it feels like we don't have the right model.   Maybe the queue should have a reference to the language server and it can directly tell it to shutdown, though maybe that should be an impl detail of the queue instead of a public api?